samza-core/src/test/java/org/apache/samza/zk/TestZkUtils.java

```diff
@@ -72,34 +84,91 @@ public static void teardown() {
 
   @Test
   public void testRegisterProcessorId() {
-    ZkUtils utils = new ZkUtils(
-        KEY_BUILDER,
-        zkClient,
-        SESSION_TIMEOUT_MS);
-    utils.connect();
-    String assignedPath = utils.registerProcessorAndGetId("0.0.0.0");
+    String assignedPath = zkUtils.registerProcessorAndGetId("0.0.0.0");
     Assert.assertTrue(assignedPath.startsWith(KEY_BUILDER.getProcessorsPath()));
 
     // Calling registerProcessorId again should return the same ephemeralPath as long as the session is valid
-    Assert.assertTrue(utils.registerProcessorAndGetId("0.0.0.0").equals(assignedPath));
+    Assert.assertTrue(zkUtils.registerProcessorAndGetId("0.0.0.0").equals(assignedPath));
 
-    utils.close();
   }
 
   @Test
   public void testGetActiveProcessors() {
-    ZkUtils utils = new ZkUtils(
-        KEY_BUILDER,
-        zkClient,
-        SESSION_TIMEOUT_MS);
-    utils.connect();
+    Assert.assertEquals(0, zkUtils.getSortedActiveProcessors().size());
+    zkUtils.registerProcessorAndGetId("processorData");
 
-    Assert.assertEquals(0, utils.getSortedActiveProcessors().size());
-    utils.registerProcessorAndGetId("processorData");
+    Assert.assertEquals(1, zkUtils.getSortedActiveProcessors().size());
 
-    Assert.assertEquals(1, utils.getSortedActiveProcessors().size());
+  }
 
-    utils.close();
+  @Test
+  public void testSubscribeToJobModelVersionChange() {
+
+    ZkKeyBuilder keyBuilder = new ZkKeyBuilder("test");
+    String root = keyBuilder.getRootPath();
+    zkClient.deleteRecursive(root);
+
+    class Result {
+      String res = "";
+      public String getRes() {
+        return res;
+      }
+      public void updateRes(String newRes) {
+        res = newRes;
+      }
+    }
+
+    Assert.assertFalse(zkUtils.exists(root));
+
+    // create the paths
+    zkUtils.makeSurePersistentPathsExists(
+        new String[]{root, keyBuilder.getJobModelVersionPath(), keyBuilder.getProcessorsPath()});
+    Assert.assertTrue(zkUtils.exists(root));
+    Assert.assertTrue(zkUtils.exists(keyBuilder.getJobModelVersionPath()));
+    Assert.assertTrue(zkUtils.exists(keyBuilder.getProcessorsPath()));
+
+    final Result res = new Result();
+    // define the callback
+    IZkDataListener dataListener = new IZkDataListener() {
```

Comment:
I am not sure if it can be done with mock. These are async calls, so we NEED the callback to fire to make sure it works.